### PR TITLE
Glancing Blow DMG and chance 

### LINKED ISSUE #10: Glancing Blow DMG
## Description

Add the damage modifier for melee glancing blows.

### PR DESCRIPTION
**Pull request by Norderud**
Added  glancing blow chance and damage in the melee section.
image

Used ChatGPT for translations, not able to confirm if correct or not.

closes #10

### CODE REVIEW

**Comment by anon1231823** [comment]
Is glancing blow a community created term? Or does Blizzard use this name in like wowhead that can be searched?

File: Modules/Config/MeleeSection.lua [hunk 1]
```diff
@@ -235,11 +235,96 @@ function _Config:LoadMeleeSection()
...
+                type = "group",
+                order = 6,
+                inline = true,
+                name = function() return i18n("Melee Glance Values") end,
+                args = {
+                    meleeGlance = {
+                        type = "toggle",
+                        order = 1,
+                        name = function() return i18n("Glancing Blow Chance") end,
+                        desc = function() return i18n("Shows/Hides the melee miss chance against enemies on the same level.") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:263]:
```suggestion
                        desc = function() return i18n("Shows/Hides the glancing chance against enemies on same level.") end,
```

File: Modules/Config/MeleeSection.lua [hunk 2]
```diff
@@ -235,11 +235,96 @@ function _Config:LoadMeleeSection()
...
+                        set = function (_, value)
+                            ExtendedCharacterStats.profile.melee.glance.sameLevel.display = value
+                            Stats:RebuildStatInfos()
+                        end,
+                    },
+                    meleeGlanceBoss = {
+                        type = "toggle",
+                        order = 2,
+                        name = function() return i18n("Glancing Blow Chance Boss") end,
+                        desc = function() return i18n("Shows/Hides the melee miss chance against boss enemies (+3 Level).") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:279]:
```suggestion
                        desc = function() return i18n("Shows/Hides the glancing chance against boss enemies (+3 Level).") end,
```

File: Modules/Data/Melee.lua [hunk 3]
```diff
@@ -176,6 +176,43 @@ function Data:MeleeHitMissChanceBossLevel()
...
+    return Data:GlanceDamageByLevel(3)
+end
+---@return string
+function Data:GlanceDamageByLevel(level)
+    local mainBase, mainMod, _, _ = UnitAttackBothHands("player")
+    local playerLevel = UnitLevel("player")
+    local enemyDefenseValue = (playerLevel + level) * 5
+
+    local glancePenalty
+    glancePenalty = DataUtils:GetGlancingDamage(mainBase + mainMod, enemyDefenseValue)
```
Comment (BreakBB) [inline Modules/Data/Melee.lua:212]:
```suggestion
    local glancePenalty = DataUtils:GetGlancingDamage(mainBase + mainMod, enemyDefenseValue)
```

File: Modules/Config/MeleeSection.lua [hunk 4]
```diff
@@ -235,11 +235,96 @@ function _Config:LoadMeleeSection()
...
+                        get = function () return ExtendedCharacterStats.profile.melee.glance.damageSameLevel.display; end,
+                        set = function (_, value)
+                            ExtendedCharacterStats.profile.melee.glance.damageSameLevel.display = value
+                            Stats:RebuildStatInfos()
+                        end,
+                    },
+                    meleeGlanceDamageBoss = {
+                        type = "toggle",
+                        order = 4,
+                        name = function() return i18n("Glancing Blow Damage") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:310]:
```suggestion
                        name = function() return i18n("Glancing Blow Damage Boss") end,
```

File: Modules/Config/MeleeSection.lua [hunk 5]
```diff
@@ -235,11 +235,96 @@ function _Config:LoadMeleeSection()
...
+                        set = function (_, value)
+                            ExtendedCharacterStats.profile.melee.glance.bossLevel.display = value
+                            Stats:RebuildStatInfos()
+                        end,
+                    },
+                    meleeGlanceDamage = {
+                        type = "toggle",
+                        order = 3,
+                        name = function() return i18n("Glancing Blow Damage") end,
+                        desc = function() return i18n("Shows/Hides the glancing dmg mod % against enemies on the same level.") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:295]: This does not match the wording in `MeleeConfigTranslations.lua`. How about:

> Shows/Hides the damage reduction against enemies on the same level.

File: Modules/Config/MeleeSection.lua [hunk 6]
```diff
@@ -235,11 +235,96 @@ function _Config:LoadMeleeSection()
...
+                        set = function (_, value)
+                            ExtendedCharacterStats.profile.melee.glance.damageSameLevel.display = value
+                            Stats:RebuildStatInfos()
+                        end,
+                    },
+                    meleeGlanceDamageBoss = {
+                        type = "toggle",
+                        order = 4,
+                        name = function() return i18n("Glancing Blow Damage") end,
+                        desc = function() return i18n("Shows/Hides the glancing dmg mod % against enemies on the same level.") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:311]: Same as with the other, the wording does not match. Maybe this fits better:

> Shows/Hides the damage reduction against enemies on boss level (+3 Level).

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 7]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
         ["esES"] = "Muestra/oculta la probabilidad de fallo contra jefes (nivel +3)",
         ["esMX"] = "Muestra/oculta la probabilidad de fallo contra jefes (nivel +3)"
     },
+    ["Glancing Blow Chance"] = {
+        ["enUS"] = "Glancing Blow Chance",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:240]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 8]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["deDE"] = "Streifschlagchance",
+        ["frFR"] = "Chance de coup éraflant",
+        ["zhCN"] = "偏斜打击几率",
+        ["ruRU"] = "Шанс скользящего удара",
+        ["esES"] = "Probabilidad de golpe de refilón",
+        ["esMX"] = "Probabilidad de golpe de refilón"
+    },
+    ["Shows/Hides the glancing chance against enemies on same level."] = {
+        ["enUS"] = "Shows/Hides the glancing chance against enemies on same level.",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:249]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 9]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["deDE"] = "Zeigt/Verbirgt die Streifschlagchance gegenüber Gegnern auf gleichem Niveau.",
+        ["frFR"] = "Affiche/Masque la chance de coup éraflant contre les ennemis du même niveau.",
+        ["zhCN"] = "显示/隐藏与同等级敌人的偏斜打击几率。",
+        ["ruRU"] = "Показывает/Скрывает шанс скользящего удара против врагов того же уровня.",
+        ["esES"] = "Muestra/Oculta la probabilidad de golpe de refilón contra enemigos del mismo nivel.",
+        ["esMX"] = "Muestra/Oculta la probabilidad de golpe de refilón contra enemigos del mismo nivel."
+    },
+    ["Glancing Blow Chance Boss"] = {
+        ["enUS"] = "Glancing Blow Chance Boss",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:258]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 10]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["deDE"] = "Streifschlagchance Boss",
+        ["frFR"] = "Chance de coup éraflant sur le boss",
+        ["zhCN"] = "对Boss的偏斜打击几率",
+        ["ruRU"] = "Шанс скользящего удара по боссу",
+        ["esES"] = "Probabilidad de golpe de refilón contra jefes",
+        ["esMX"] = "Probabilidad de golpe de refilón contra jefes"
+    },
+    ["Shows/Hides the glancing chance against boss enemies (+3 Level)."] = {
+        ["enUS"] = "Shows/Hides the glancing chance against boss enemies (+3 Level).",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:267]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 11]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["deDE"] = "Zeigt/Verbirgt die Streifschlagchance gegenüber Bossgegnern (+3 Stufen).",
+        ["frFR"] = "Affiche/Masque la chance de coup éraflant contre les boss (+3 Niveaux).",
+        ["zhCN"] = "显示/隐藏对Boss敌人的偏斜打击几率（+3级）。",
+        ["ruRU"] = "Показывает/Скрывает шанс скользящего удара против боссов (+3 Уровня).",
+        ["esES"] = "Muestra/Oculta la probabilidad de golpe de refilón contra jefes (+3 Niveles).",
+        ["esMX"] = "Muestra/Oculta la probabilidad de golpe de refilón contra jefes (+3 Niveles)."
+    },
+    ["Glancing Blow Damage"] = {
+        ["enUS"] = "Glancing Blow Damage",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:276]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 12]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+    },
+    ["Shows/Hides the glancing damage against enemies on same level"] = {
+        ["enUS"] = "Shows/Hides the glancing damage against enemies on same level",
+        ["deDE"] = "Zeigt/Verbirgt den Streifschlagschaden gegenüber Gegnern auf gleichem Niveau",
+        ["frFR"] = "Affiche/Masque les dégâts du coup éraflant contre les ennemis du même niveau",
+        ["zhCN"] = "显示/隐藏对同等级敌人的偏斜打击伤害",
+        ["ruRU"] = "Показывает/Скрывает урон от скользящего удара против врагов того же уровня",
+        ["esES"] = "Muestra/Oculta el daño de golpe de refilón contra enemigos del mismo nivel",
+        ["esMX"] = "Muestra/Oculta el daño de golpe de refilón contra enemigos del mismo nivel"
+    },
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:292]: These needs to be updated as well, after the translation is changed.

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 13]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+    ["Shows/Hides the glancing damage against boss enemies (+3 Level)."] = {
+        ["enUS"] = "Shows/Hides the glancing damage against boss enemies (+3 Level).",
+        ["deDE"] = "Zeigt/Verbirgt den Streifschlagschaden gegenüber Bossgegnern (+3 Stufen).",
+        ["frFR"] = "Affiche/Masque les dégâts du coup éraflant contre les boss (+3 Niveaux).",
+        ["zhCN"] = "显示/隐藏对Boss敌人的偏斜打击伤害（+3级）。",
+        ["ruRU"] = "Показывает/Скрывает урон от скользящего удара против боссов (+3 Уровня).",
+        ["esES"] = "Muestra/Oculta el daño de golpe de refilón contra jefes (+3 Niveles).",
+        ["esMX"] = "Muestra/Oculta el daño de golpe de refilón contra jefes (+3 Niveles)."
+    },
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:310]: Same as with the other.

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 14]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["enUS"] = "Shows/Hides the glancing damage against enemies on same level",
+        ["deDE"] = "Zeigt/Verbirgt den Streifschlagschaden gegenüber Gegnern auf gleichem Niveau",
+        ["frFR"] = "Affiche/Masque les dégâts du coup éraflant contre les ennemis du même niveau",
+        ["zhCN"] = "显示/隐藏对同等级敌人的偏斜打击伤害",
+        ["ruRU"] = "Показывает/Скрывает урон от скользящего удара против врагов того же уровня",
+        ["esES"] = "Muestra/Oculta el daño de golpe de refilón contra enemigos del mismo nivel",
+        ["esMX"] = "Muestra/Oculta el daño de golpe de refilón contra enemigos del mismo nivel"
+    },
+    ["Glancing Blow Damage Boss"] = {
+        ["enUS"] = "Glancing Blow Damage Boss",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:294]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 15]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["enUS"] = "Shows/Hides the glancing damage against boss enemies (+3 Level).",
+        ["deDE"] = "Zeigt/Verbirgt den Streifschlagschaden gegenüber Bossgegnern (+3 Stufen).",
+        ["frFR"] = "Affiche/Masque les dégâts du coup éraflant contre les boss (+3 Niveaux).",
+        ["zhCN"] = "显示/隐藏对Boss敌人的偏斜打击伤害（+3级）。",
+        ["ruRU"] = "Показывает/Скрывает урон от скользящего удара против боссов (+3 Уровня).",
+        ["esES"] = "Muestra/Oculta el daño de golpe de refilón contra jefes (+3 Niveles).",
+        ["esMX"] = "Muestra/Oculta el daño de golpe de refilón contra jefes (+3 Niveles)."
+    },
+    ["Melee Glance Values"] = {
+        ["enUS"] = "Melee Glance Values",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:312]:
```suggestion
        ["enUS"] = true,
```

File: Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua [hunk 16]
```diff
@@ -236,6 +236,96 @@ local meleeConfigTranslations = {
...
+        ["deDE"] = "Nahkampf-Streifwerte",
+        ["frFR"] = "Valeurs de coup éraflant en mêlée",
+        ["zhCN"] = "近战偏斜值",
+        ["ruRU"] = "Значения скользящих ударов в ближнем бою",
+        ["esES"] = "Valores de refilón en combate cuerpo a cuerpo",
+        ["esMX"] = "Valores de refilón en combate cuerpo a cuerpo"
+    },
+    ["Shows/Hides all glancing blow stats"] = {
+        ["enUS"] = "Shows/Hides all glancing blow stats",
```
Comment (BreakBB) [inline Modules/i18n/translations/ConfigTranslations/MeleeConfigTranslations.lua:321]:
```suggestion
        ["enUS"] = true,
```

**Review by BreakBB — CHANGES_REQUESTED** [review]
Hey @puggan 

first of all: Thank you very much for looking into adding these stats to ECS! I really appreciate any help I get on this project.

I left a few comments and suggestions in the code itself. On top of that, the calculation for Glancing Blows which you added, is the one for Era/HC/SoD, therefore these values should be hidden for everything else.

The settings can be hidden by adding the `hidden` field:
```lua
                        hidden = function()
                            return (not ECS.IsClassic)
                        end,
```

To not add the values itself, checkout `Stats.lua` -> `_CreateStatInfos` in which we already use `ECS.IsWotlk` to hide specific stats on non-WotLK clients. You can adjust these to also check for `ECS.IsClassic` to then conditionally add the Glancing Blow stats.

Let me know if you want to tackle these, otherwise I'll look into adjusting the PR to get it out and released 💪🏻

**Comment by BreakBB** [comment]
Sounds good @Norderud ! Let me know, when you are done or if you need further help/ideas 👍🏻

File: Modules/Config/MeleeSection.lua [hunk 17]
```diff
@@ -235,11 +235,102 @@ function _Config:LoadMeleeSection()
...
+                        set = function (_, value)
+                            ExtendedCharacterStats.profile.melee.glance.bossLevel.display = value
+                            Stats:RebuildStatInfos()
+                        end,
+                    },
+                    meleeGlanceDamage = {
+                        type = "toggle",
+                        order = 3,
+                        name = function() return i18n("Glancing Blow Damage") end,
+                        desc = function() return i18n("Shows/Hides the damage reduction against enemies on the same level.") end,
```
Comment (BreakBB) [inline Modules/Config/MeleeSection.lua:301]: The translation does not match with this. I like the wording though and think it should be the same for `meleeGlanceDamageBoss` a few lines below

**Review by BreakBB — CHANGES_REQUESTED** [review]

**Review by BreakBB — APPROVED** [review]

**Comment by BreakBB** [comment]
Thank you @Norderud for continuing to work on this! Looking good to me. I'll see to ship it in a release very soon.

**Pull request merged**